### PR TITLE
D-SCRIPT 0.2.0 -- Topsy-Turvy

### PR DESCRIPTION
**Pull request by samsledje**
- Integrate Topsy-Turvy to allow for top-down supervision
- Use utils.log function across all commands
- Speed up loading embeddings into memory using parallel processing
- Update fasta parse and write to use BioPython SeqIO (better error checking)
- More comprehensive test suite for main commands

### CODE REVIEW

File: dscript/commands/predict.py [hunk 1]
```diff
@@ -98,48 +96,43 @@ def main(args):
...
-        log(f"Pairs File {csvPath} not found", file=logFile)
+        log(f"Pairs File {csvPath} not found", file=logFile, print_also=True)
         logFile.close()
         sys.exit(1)
 
     # Load Sequences or Embeddings
     if embPath is None:
         try:
-            names, seqs = parse(open(seqPath, "r"))
+            names, seqs = parse(seqPath, "r")
```
Comment (rs239) [inline dscript/commands/predict.py:116]: does this need to be parse(open(seqPath...)??

File: dscript/commands/train.py [hunk 2]
```diff
@@ -150,9 +155,31 @@ def add_args(parser):
...
+        type=float,
+        default=0.2,
+        help="weight on the GLIDER accuracy objective (default: 0.2)",
+    )
+    train_grp.add_argument(
+        "--glider-thresh",
+        dest="glider_thresh",
+        type=float,
+        default=0.925,
+        help="proportion of GLIDER scores treated as positive edges (0 < gt < 1) (default: 0.925)",
```
Comment (rs239) [inline dscript/commands/train.py:177]: change ""proportion of GLIDER scores treated as positive edges (0 < gt < 1) (default: 0.925)""
to 
""threshold beyond which GLIDER scores treated as positive edges (0 < gt < 1) (default: 0.925)"

File: dscript/fasta.py [hunk 3]
```diff
@@ -1,51 +1,16 @@
...
-            name = line[1:]
-            sequence = []
-        else:
-            sequence.append(line.upper())
-    if name is not None:
-        names.append(name)
-        sequences.append(empty.join(sequence))
+
+    for record in SeqIO.parse(f, "fasta"):
+        names.append(record.name)
```
Comment (rs239) [inline dscript/fasta.py:12]: In previous versions we would take the full description line as the sequence ID. We should take just the first word (i.e. split on space). Not sure of record.name does that?

File: dscript/commands/predict.py (hunk 1 above)
Comment (samsledje) [reply dscript/commands/predict.py:116]: Updated `parse` command in commit dfbc81829f9e4f56a2be93fcdf8d7530687d1ba9 to use BioPython SeqIO, which can take a file path or an open file object

File: dscript/fasta.py (hunk 3 above)
Comment (samsledje) [reply dscript/fasta.py:12]: SeqIO records parse everything before the first space as the record name, and the full line as the record description -- this should be giving the desired behavior

File: dscript/commands/train.py (hunk 2 above)
Comment (samsledje) [reply dscript/commands/train.py]: Updated in 1b598e47620f313ebba9610154a87ee5785bf16d

**Review by rs239 — COMMENTED** [review]
nicely done!

**Review by rs239 — APPROVED** [review]

**Pull request merged**